packages/financial-templates-lib/src/clients/InsuredBridgeL1Client.ts

```diff
@@ -120,6 +129,68 @@ export class InsuredBridgeL1Client {
+    this._throwIfNotInitialized();
+    const l1TokenNormalized = toChecksumAddress(l1Token);
+
+    if (
+      !this.updatedRateModelEventsForToken[l1TokenNormalized] ||
+      this.updatedRateModelEventsForToken[l1TokenNormalized].length === 0
+    )
+      throw new Error(`No updated rate model events for L1 token: ${l1TokenNormalized}`);
+
+    if (!blockNumber) {
+      // If block number is undefined, return latest updated rate model.
+      return this.updatedRateModelEventsForToken[l1TokenNormalized].slice(-1)[0].rateModel;
+    } else {
+      const firstEventBlockNumber = this.updatedRateModelEventsForToken[l1TokenNormalized][0].blockNumber;
+      if (blockNumber < firstEventBlockNumber) {
+        throw new Error(
+          `Block number #${blockNumber} is before first UpdatedRateModel event block ${firstEventBlockNumber}`
+        );
+      }
+
+      // We're looking for the latest rate model update that occurred at or before the block number.
+      // Rate model events are inserted into the array from oldest at index 0 to newest at index length-1, so we'll
+      // reverse the array so it goes from newest at index 0 to oldest at index length-1, and then find the first event
+      // who's block number is less than or equal to the target block number.
+      const rateModel = this.updatedRateModelEventsForToken[l1TokenNormalized]
+        .slice() // reverse() modifies memory in place so create a copy first.
+        .reverse()
+        .find((event) => event.blockNumber <= blockNumber);
+
+      if (!rateModel)
+        throw new Error(`No updated rate model events before block #${blockNumber} for L1 token: ${l1TokenNormalized}`);
+      return rateModel?.rateModel;
+    }
+  }
+
+  /**
+   * @notice Return all L1 tokens that had a rate model associated with it at the block number.
+   * @param blockNumber Returns l1 tokens that were mapped to a rate model at this block height. If undefined,
+   * this function will return all L1 tokens that have a block number as of the latest block height.
+   * @returns array of L1 token addresses.
+   */
+  getL1TokensFromRateModel(blockNumber: number | undefined = undefined): string[] {
+    this._throwIfNotInitialized();
+
+    return Object.keys(this.updatedRateModelEventsForToken)
+      .map((l1Token) => {
+        const l1TokenNormalized = toChecksumAddress(l1Token);
+
+        // Check that there is at least one UpdatedRateModel event before the provided block number, otherwise
+        // this L1 token didn't exist in the RateModel at the block height and we shouldn't include it in the returned
+        // array.
+        if (
+          !blockNumber ||
+          this.updatedRateModelEventsForToken[l1TokenNormalized].find((event) => event.blockNumber <= blockNumber)
+        )
+          return toChecksumAddress(l1Token);
+        else return null;
+      })
+      .filter((x) => x !== null) as string[];
```

Comment:
nit: you should be able to use [this function](https://github.com/UMAprotocol/protocol/blob/3955d80038df1c54663a59b44d6e23bd09c7dcdc/packages/financial-templates-lib/src/types.ts#L12-L14) to force typescript to understand that the output will be a string.

Example of how this is used:
```ts
// can be imported from ../types
function isDefined<T>(val: T | undefined | null): val is T {
  return val !== undefined && val !== null;
}
....

.filter(isDefined);
```

If this all works correctly, no cast should be necessary. See https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates for more details on this concept.

Reply:
Oh very nice